compiler-access/src/main/scala/org/alephium/ralph/lsp/access/compiler/ast/node/Node.scala

```diff
@@ -0,0 +1,94 @@
+package org.alephium.ralph.lsp.access.compiler.ast.node
+
+object Node {
+  /** Create a [[Node]] with no children */
+  @inline def apply[A](data: A): Node[A] =
+    new Node(data, List.empty)
+
+  /** Create a [[Node]] with children */
+  @inline def apply[A](data: A, children: Seq[Node[A]]): Node[A] = {
+    val parent = new Node(data, children)
+    // set the parent for each child node to allow traversing up.
+    children foreach (_._parent = Some(parent))
+    parent
+  }
+}
+
+/**
+ * A [[Node]] represents a single position within a tree.
+ *
+ * Each node allows tree traversal in both forward and backward directions
+ * using the functions like [[walkDown]], [[walkParents]] and others.
+ *
+ * @param data     The data stored in this node.
+ * @param children This node's child nodes.
+ * @tparam A Data type.
+ */
+case class Node[A] private(data: A,
+                           children: Seq[Node[A]]) { self =>
+  private var _parent: Option[Node[A]] =
+    None
+
+  def parent: Option[Node[A]] =
+    _parent
+
+  def headOption: Option[Node[A]] =
+    parent
+      .iterator
+      .flatMap(_.walkParents)
+      .foldLeft(Option.empty[Node[A]]) {
+        case (_, next) =>
+          Some(next)
+      }
+
+  /** Walk down from current node reaching all it's children and grand-children */
+  def walkDown: Iterator[Node[A]] =
+    new Iterator[Node[A]] {
+      private val iter =
+        Iterator(self) ++
+          children
+            .iterator
+            .flatMap(_.walkDown)
```

Comment:
is there a risk of infinite walkdown, if someone set a parent as a children?

Reply:
I don't think so. There is no [constructor](https://github.com/alephium/ralph-lsp/pull/121/files#diff-fa0e48d764d8246a9a4f74af69d0f9d4b9ecf9c64ab2feed17e059c9730bf664R5-R14) that would allow that to happen. Unless you've found a way?